src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoOfT.WriteHelpers.cs

```diff
@@ -0,0 +1,336 @@
+                rootValue is not null &&
+                Options.TryGetPolymorphicTypeInfoForRootType(rootValue, out JsonTypeInfo? polymorphicTypeInfo))
+            {
+                Debug.Assert(typeof(T) == typeof(object));
+                polymorphicTypeInfo.SerializeAsObject(utf8Json, rootValue, isInvokedByPolymorphicConverter: true);
+            }
+            else
+            {
+                bool isFinalBlock;
+                WriteStack state = default;
+                state.Initialize(this,
+                    rootValueBoxed,
+                    isInvokedByPolymorphicConverter,
+                    supportContinuation: true,
+                    supportAsync: false);
+
+                using var bufferWriter = new PooledByteBufferWriter(Options.DefaultBufferSize);
+                using var writer = new Utf8JsonWriter(bufferWriter, Options.GetWriterOptions());
+
+                try
+                {
+                    do
+                    {
+                        state.FlushThreshold = (int)(bufferWriter.Capacity * JsonSerializer.FlushThreshold);
+
+                        isFinalBlock = EffectiveConverter.WriteCore(writer, rootValue!, Options, ref state);
+                        writer.Flush();
+
+                        bufferWriter.WriteToStream(utf8Json);
+                        bufferWriter.Clear();
+
+                        Debug.Assert(state.PendingTask == null);
+                    } while (!isFinalBlock);
+                }
+                finally
+                {
+                    if (CanUseSerializeHandler)
+                    {
+                        OnRootLevelAsyncSerializationCompleted(writer.BytesCommitted + writer.BytesPending);
+                    }
+                }
+            }
+        }
+
+        internal sealed override void SerializeAsObject(Utf8JsonWriter writer, object? rootValue, bool isInvokedByPolymorphicConverter = false)
+            => Serialize(writer, UnboxValue(rootValue), rootValue, isInvokedByPolymorphicConverter);
+
+        internal sealed override Task SerializeAsObjectAsync(Stream utf8Json, object? rootValue, CancellationToken cancellationToken, bool isInvokedByPolymorphicConverter = false)
+            => SerializeAsync(utf8Json, UnboxValue(rootValue), cancellationToken, rootValue, isInvokedByPolymorphicConverter);
+
+        internal sealed override void SerializeAsObject(Stream utf8Json, object? rootValue, bool isInvokedByPolymorphicConverter = false)
+            => Serialize(utf8Json, UnboxValue(rootValue), rootValue, isInvokedByPolymorphicConverter);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private T? UnboxValue(object? value)
+        {
+            if (
+#if NETCOREAPP
+                // Treated as a constant by recent versions of the JIT.
+                typeof(T).IsValueType &&
+#else
+                Type.IsValueType &&
+#endif
+                default(T) is not null && value is null)
```

Comment:
I'm not sure I understand the point of the `default(T) is not null` check. If we know that `T` is a ValueType, when will `default(T) is not null` ever be `false`? It seems like a redundant check to me.

Reply:
Oh - it's for `Nullable<T>` ... right?

Reply:
It can only be true if your T is `Nullable<T>`, which is the case we're meant to catch here.

Reply:
FWIW this is pre-existing logic that got moved from `JsonConverterOfT.WriteCore.cs`

Reply:
Hmm... actually you're right, that check should have been `default(T) is null && value is null`. Surprised this isn't caught by any tests.

Reply:
I _think_ the code is correct, now that I realized it was for checking for Nullable.  When T is Nullable, `typeof(T).IsValueType` will be `true` and `default(T) is not null` will be `false`. So the exception won't be thrown in the case of `value` being `null`. Because it is fine for `Nullable<T>` to have a `null` value.

I guess my one suggestion would be to throw a comment here to help the next person realize faster that this check is for Nullable.

Reply:
> I guess my one suggestion would be to throw a comment here to help the next person realize faster that this check is for Nullable.

There is a comment about that in line 291.